sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/MapStateImplWithTTL.scala

```diff
@@ -0,0 +1,280 @@
+    store.remove(compositeKey, stateName)
+  }
+
+  /** Remove this state. */
+  override def clear(): Unit = {
+    store.removeColFamilyIfExists(stateName)
+    createColumnFamily()
+    clearTTLState()
+  }
+
+  /**
+   * Clears the user state associated with this grouping key
+   * if it has expired. This function is called by Spark to perform
+   * cleanup at the end of transformWithState processing.
+   *
+   * Spark uses a secondary index to determine if the user state for
+   * this grouping key has expired. However, its possible that the user
+   * has updated the TTL and secondary index is out of date. Implementations
+   * must validate that the user State has actually expired before cleanup based
+   * on their own State data.
+   *
+   * @param groupingKey grouping key for which cleanup should be performed.
+   * @param userKey     user key for which cleanup should be performed.
+   */
+  override def clearIfExpired(groupingKey: Array[Byte], userKey: Array[Byte]): Long = {
+    val encodedCompositeKey = stateTypesEncoder.encodeCompositeKey(groupingKey, userKey)
+    val retRow = store.get(encodedCompositeKey, stateName)
+    var numRemovedElements = 0L
+    if (retRow != null) {
+      if (stateTypesEncoder.isExpired(retRow, batchTimestampMs)) {
+        store.remove(encodedCompositeKey, stateName)
+        numRemovedElements += 1
+      }
+    }
+    numRemovedElements
+  }
+
+  /*
+   * Internal methods to probe state for testing. The below methods exist for unit tests
+   * to read the state ttl values, and ensure that values are persisted correctly in
+   * the underlying state store.
+   */
+
+  /**
+   * Retrieves the value from State even if its expired. This method is used
+   * in tests to read the state store value, and ensure if its cleaned up at the
+   * end of the micro-batch.
+   */
+  private[sql] def getWithoutEnforcingTTL(userKey: K): Option[V] = {
+    val encodedCompositeKey = stateTypesEncoder.encodeCompositeKey(userKey)
+    val retRow = store.get(encodedCompositeKey, stateName)
+
+    if (retRow != null) {
+      val resState = stateTypesEncoder.decodeValue(retRow)
+      Some(resState)
+    } else {
+      None
+    }
+  }
+
+  /**
+   * Read the ttl value associated with the grouping and user key.
+   */
+  private[sql] def getTTLValue(userKey: K): Option[(V, Long)] = {
+    val encodedCompositeKey = stateTypesEncoder.encodeCompositeKey(userKey)
+    val retRow = store.get(encodedCompositeKey, stateName)
+
+    // if the returned row is not null, we want to return the value associated with the
+    // ttlExpiration
+    Option(retRow).flatMap { row =>
+      val ttlExpiration = stateTypesEncoder.decodeTtlExpirationMs(row)
+      ttlExpiration.map(expiration => (stateTypesEncoder.decodeValue(row), expiration))
+    }
+  }
+
+  /**
+   * Get all ttl values stored in ttl state for current implicit
+   * grouping key.
+   */
+  private[sql] def getValuesInTTLState(): Iterator[Long] = {
```

Comment:
Do we need to implement both methods? This seems to be just `getKeyValuesInTTLState().map(_._2)`, and since it's only used for testing we don't expect these methods to be optimized at best.

Reply:
Good point, removed this method.